src/functional/iterator/transforms/cse.py

```diff
@@ -56,16 +56,19 @@ def visit_Lambda(
 
     def visit_FunCall(
         self,
-        node: ir.Lambda,
+        node: ir.FunCall,
         *,
         subexprs: dict[ir.Node, tuple[list[int], Optional[ir.Node]]],
         refs: ChainMap[str, bool],
         parent: Optional[ir.Node],
         collector_stack: list[bool],
     ) -> None:
-        # do not collect (and thus deduplicate in CSE) shift(offsets…) calls. Node must still be
-        #  visited, to ensure symbol dependencies are recognized correctly.
-        allow_collection = node.fun != ir.SymRef(id="shift")
+        allow_collection = node.fun != ir.SymRef(id="shift") and not (
+            isinstance(
+                node.fun, ir.FunCall
+            )  # TODO: want to remove the guard after the `or` (gtfn_im backend)
```

Comment:
yes, this pertains to the second bullet under "limitations" in the description to this PR. It reads:

"Shift constructs are not lifted out of the arguments of reductions to avoid situations where there are no partial shifts in the argument list of reductions"

So basically we want to avoid this:
``` {dwdx, dwdy} ← (
      λ(w, geofac_grg_x, geofac_grg_y) → 
        {reduce(λ(acc, geofac_grg_x__0, w__1) → acc + geofac_grg_x__0 × w__1, 0)(geofac_grg_x, ⟪C2E2COₒ⟫(w)),
         reduce(λ(acc_, geofac_grg_y__0, w__1_) → acc_ + geofac_grg_y__0 × w__1_, 0)(
           geofac_grg_y, ⟪C2E2COₒ⟫(w)
         )}
    )(w, geofac_grg_x, geofac_grg_y) @ u⟨ horizontal: [horizontal_start, horizontal_end), K: [vertical_start, vertical_end) ⟩;```

turning into this:
```  {dwdx, dwdy} ← (
      λ(w, geofac_grg_x, geofac_grg_y) → 
        (λ(_cs_1) → 
           {reduce(λ(acc, geofac_grg_x__0, w__1) → acc + geofac_grg_x__0 × w__1, 0)(geofac_grg_x, _cs_1),
            reduce(λ(acc_, geofac_grg_y__0, w__1_) → acc_ + geofac_grg_y__0 × w__1_, 0)(geofac_grg_y, _cs_1)})(
          ⟪C2E2COₒ⟫(w)
        )
    )(w, geofac_grg_x, geofac_grg_y) @ u⟨ horizontal: [horizontal_start, horizontal_end), K: [vertical_start, vertical_end) ⟩;```